### DISCUSSION TITLE
A git buildpack

### DISCUSSION THREAD

**Discussion by genevieve** [paketo-buildpacks/feedback#4] [Ideas]
While trying to build a go app with the go buildpack, we hit a snag trying to inject the git sha at build-time. In a dockerfile it might look something like:
```
RUN GIT_COMMIT=$(git rev-list -1 HEAD) && \
  go build -ldflags "-X main.GitCommit=$GIT_COMMIT"
```

The go buildpack supports buildtime flags for the `-ldflags ...` part but there isn't an easy way to extract the first part where the sha is deduced. Would this be something y'all would be interesting in creating?

**Comment by ryanmoran** [comment]
I think this make sense to support, but maybe not directly in the `go-build` buildpack. Instead, I could imagine a `git` buildpack that will optionally participate in a build and provide `git` metadata as build-time environment variables, making them accessible to a number of different buildpacks. This would make this type of feature accessible across all of the language-families.

**Reply by ForestEckhardt** [reply]
Could you provide me some other uses cases for this kind of buildpack? I am not opposed to the idea I am just struggling to think how else this might be utilized.

**Reply by genevieve** [reply]
Besides the usecase for `go build`, I've also seen multiple Rails applications that want a "version" (in non-release cases, they want the git sha) for different purposes at runtime.

**Comment by arjun024** [comment]
Just curious - does your build process create commits? Wondering why injecting the value before build-time wouldn't be sufficient? like,
`pack build app1 -p <path> --env BP_GO_BUILD_FLAGS="-ldflags=\"-X main.GitCommit=$(git -C <path> rev-list -1 HEAD)\""`

**Reply by genevieve** [reply]
Hey! Our pack build command is in the circleci orb and used by more apps than the golang one that I'm working on so it's not easy to introduce a custom flag arg. Additionally, I'm not sure this would work when we move to kpack and are using the image/build resources with env var configuration.

**Reply by arjun024** [reply]
Thanks for the background. In case you're interested to make a quick buildpack to provide the variable, you should be able to do so in a few lines since `git` is provided by the paketo full and base images.
```
‣ cat bin/detect
#!/usr/bin/env bash

‣ cat bin/build
#!/usr/bin/env bash
mkdir -p "${1}/layer1/env.build"
echo -n "build = true" > "${1}/layer1.toml"
git rev-list -1 HEAD | tr -d "\n" > "${1}/layer1/env.build/GIT_COMMIT.default"

‣ cat buildpack.toml
api = "0.4"
[buildpack]
  id = "gitvar"
  version = "0.1"
[[stacks]]
  id = "io.buildpacks.stacks.bionic"
```

**Comment by genevieve** [comment]
RFC here: https://github.com/paketo-buildpacks/rfcs/pull/39
